packages/@react-aria/virtualizer/src/ScrollView.tsx

```diff
@@ -147,32 +147,38 @@ function ScrollView(props: ScrollViewProps, ref: RefObject<HTMLDivElement>) {
     if (state.width !== w || state.height !== h) {
       state.width = w;
       state.height = h;
-      onVisibleRectChange(new Rect(state.scrollLeft, state.scrollTop, w, h));
+      flush(() => {
+        onVisibleRectChange(new Rect(state.scrollLeft, state.scrollTop, w, h));
+      });
+
+      // If the clientWidth or clientHeight changed, scrollbars appeared or disappeared as
+      // a result of the layout update. In this case, re-layout again to account for the
+      // adjusted space. In very specific cases this might result in the scrollbars disappearing
+      // again, resulting in extra padding. We stop after a maximum of two layout passes to avoid
+      // an infinite loop. This matches how browsers behavior with native CSS grid layout.
+      if (!isTestEnv && clientWidth !== dom.clientWidth || clientHeight !== dom.clientHeight) {
+        state.width = dom.clientWidth;
+        state.height = dom.clientHeight;
+        flush(() => {
+          onVisibleRectChange(new Rect(state.scrollLeft, state.scrollTop, state.width, state.height));
+        });
+      }
     }
-  }, [onVisibleRectChange, ref, state, sizeToFit, contentSize]);
+  });
 
   useLayoutEffect(() => {
-    updateSize();
+    // React doesn't allow flushSync inside effects so pass an identity function instead.
+    // This only happens on initial render. The resize observer will also call updateSize
+    // once it initializes, but we need earlier initialization in a layout effect to avoid
+    // a flash of missing content.
+    updateSize(fn => fn());
   }, [updateSize]);
-  let raf = useRef<ReturnType<typeof requestAnimationFrame> | null>();
   let onResize = useCallback(() => {
-    if (isOldReact) {
-      raf.current ??= requestAnimationFrame(() => {
```

Comment:
We no longer need the RAF to avoid the ResizeObserver loop error. It's not possible for our layout update to cause another ResizeObserver notification because we now watch the border-box instead of the content-box, so when scrollbars appear or disappear, the ResizeObserver is not fired.